src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -0,0 +1,244 @@
+                }
+                else if (keySegment.Equals(Fingerprint.Wildcard, StringComparison.Ordinal))
+                {
+                    throw new ArgumentException("`**` is a reserved key segment. For path glob, use `./**`.");
+                }
+                else if (IsPathy(keySegment))
+                {
+                    context.Verbose($"Interpretting `{keySegment}` as a path.");
+
+                    var segment = new StringBuilder();
+                    bool foundFile = false;
+
+                    if (keySegment.Contains(';', StringComparison.Ordinal))
+                    {
+                        throw new ArgumentException("Cache key cannot contain the ';' character.");
+                    }
+
+                    string[] pathRules = keySegment.Split(new []{','}, StringSplitOptions.RemoveEmptyEntries);
+                    string rootRule = pathRules.First();
+                    if(rootRule.Length == 0 || rootRule[1] == '!')
+                    {
+                        throw new ArgumentException();
+                    }
+
+                    string workingDirectory = null;
+                    if (!IsAbsolutePath(rootRule))
+                    {
+                        workingDirectory = workingDirectoryValue;
+                    }
+
+                    string absoluteRootRule = MakePathAbsolute(workingDirectory, rootRule);
+                    context.Verbose($"Expanded include rule is `{absoluteRootRule}`.");
+                    IEnumerable<string> absoluteExcludeRules = pathRules.Skip(1).Select(r => MakePathAbsolute(workingDirectory, r.Substring(1)));
+                    Func<string,bool> filter = CreateFilter(context, workingDirectory, absoluteRootRule, absoluteExcludeRules);
+
+                    DetermineEnumeration(
+                        workingDirectory,
+                        absoluteRootRule,
+                        out string enumerateRootPath,
+                        out string enumeratePattern,
+                        out SearchOption enumerateDepth);
+
+                    context.Verbose($"Enumerating starting at root `{enumerateRootPath}` with pattern `{enumeratePattern}`.");
+                    IEnumerable<string> files = Directory.EnumerateFiles(enumerateRootPath, enumeratePattern, enumerateDepth);
+                    files = files.Where(f => filter(f)).Distinct();
+
+                    foreach(string path in files)
+                    {
+                        foundFile = true;
+
+                        using (var fs = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read))
+                        {
+                            byte[] hash = sha256.ComputeHash(fs);
+                            string displayPath = workingDirectory == null ? path : path.Substring(enumerateRootPath.Length + 1);
+                            segment.Append($"\nSHA256({displayPath})=[{fs.Length}]{hash.ToHex()}");
```

Comment:
If the developer changed their `key` to start using fully-qualified paths instead of relative paths (but the end result is the same set of files), will it change the fingerprint? I think it will, which seems like a problem.

For example:

```
key: yarn.lock
```

changes to:

```
key: $(Build.SourcesDirectory)/yarn.lock
```

Reply:
yes it will but `key: $(Build.SourcesDirectory)/yarn.lock` is bad because it can change from one private agent to another.

Reply:
Ok we will go relative to pipeline.workspace when we can